### PR TITLE
HIVE-25458: Unix_timestamp() with string input give wrong result

### PR DESCRIPTION
**Pull request by ashish-kumar-sharma**
### What changes were proposed in this pull request?

unix_timestamp() accept 4 value string/date/timestamp/timestamptz. Out of which date/timestamp/timestamptz use DateTimeFormatter.class where as string type use SimpleDateTimeformatter.class which cause difference is value.

Example -
select from_unixtime(unix_timestamp('1800-11-08 01:53:11'));
1800-11-08 01:35:15

select from_unixtime(unix_timestamp(cast('1800-11-08 01:53:11' as timestamp)));
1800-11-08 01:53:11

### Why are the changes needed?

moving string input type processing from simpledatetimeformatter to DateTimeFromatter.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UTs and QTs are added

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java [hunk 1]
```diff
@@ -20,10 +20,14 @@
 
 import java.text.ParseException;
 import java.text.SimpleDateFormat;
```
Comment (adesh-rao) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java:22]: nit: remove unused imports

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java [hunk 2]
```diff
@@ -164,15 +164,13 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
...
+          formatter = DateTimeFormatter.ofPattern(patternVal);
           lasPattern = patternVal;
         }
       }
-      try {
-        retValue.set(formatter.parse(textVal).getTime() / 1000);
-      } catch (ParseException e) {
-        return null;
-      }
+      Timestamp timestamp = new Timestamp(LocalDateTime.parse(textVal,formatter));
```
Comment (adesh-rao) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java:171]: nit: spaces after comma

File: ql/src/test/queries/clientpositive/udf5.q [hunk 3]
```diff
@@ -45,3 +45,10 @@ select from_unixtime(to_unix_timestamp(cast('0000-00-00' as date)));
 set time zone Europe/Rome;
 
 SELECT from_unixtime(1226446340), to_date(from_unixtime(1226446340)), day('2008-11-01'), month('2008-11-01'), year('2008-11-01'), day('2008-11-01 15:32:20'), month('2008-11-01 15:32:20'), year('2008-11-01 15:32:20') FROM dest1_n14;
+
+set hive.local.time.zone=Asia/Bangkok;
+
+select from_unixtime(unix_timestamp('1400-11-08 01:53:11'));
+select from_unixtime(unix_timestamp('1800-11-08 01:53:11'));
+select from_unixtime(unix_timestamp('1400-11-08 08:00:00 ICT', 'yyyy-MM-dd HH:mm:ss z'));
```
Comment (adesh-rao) [inline ql/src/test/queries/clientpositive/udf5.q:53]: Can we add a test for a different timezone too? both asia/Bangkok and ICT are GMT +7

Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/udf5.q]: done

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java (hunk 2 above)
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java]: done

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java (hunk 1 above)
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java]: done

**Review by guptanikhil007 — APPROVED** [review]

**Review by adesh-rao — APPROVED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java [hunk 4]
```diff
@@ -164,15 +164,32 @@ public Object evaluate(DeferredObject[] arguments) throws HiveException {
...
+          timestamp = new Timestamp(localDateTime);
+        } catch (DateTimeException e) {
+          try {
+            LocalDate localDate = LocalDate.parse(textVal, formatter);
+            timestamp = new Timestamp(localDate.atStartOfDay());
+          } catch (DateTimeException ex) {
+            return null;
+          }
+        }
+
```
Comment (sankarh) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java:182]: Nit: Remove extra line.

File: ql/src/test/queries/clientpositive/udf5.q [hunk 5]
```diff
@@ -9,24 +9,24 @@ SELECT from_unixtime(1226446340), to_date(from_unixtime(1226446340)), day('2008-
 SELECT from_unixtime(1226446340), to_date(from_unixtime(1226446340)), day('2008-11-01'), month('2008-11-01'), year('2008-11-01'), day('2008-11-01 15:32:20'), month('2008-11-01 15:32:20'), year('2008-11-01 15:32:20') FROM dest1_n14;
 
 EXPLAIN
-SELECT from_unixtime(unix_timestamp('2010-01-13 11:57:40', 'yyyy-MM-dd HH:mm:ss'), 'MM/dd/yy HH:mm:ss'), from_unixtime(unix_timestamp('2010-01-13 11:57:40')) from dest1_n14;
+SELECT from_unixtime(unix_timestamp('2010-01-13 11:57:40', 'uuuu-MM-dd HH:mm:ss'), 'MM/dd/yy HH:mm:ss'), from_unixtime(unix_timestamp('2010-01-13 11:57:40')) from dest1_n14;
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/udf5.q:12]: Instead of changing existing tests, can we add new ones with "uuuu"?

File: ql/src/test/queries/clientpositive/udf5.q [hunk 6]
```diff
@@ -45,3 +45,24 @@ select from_unixtime(to_unix_timestamp(cast('0000-00-00' as date)));
...
+
+select from_unixtime(unix_timestamp('1400-11-08 01:53:11'));
+select from_unixtime(unix_timestamp('1400-11-08', 'yyyy-MM-dd'));
+select from_unixtime(unix_timestamp("Wed, Jul 4, '01", "EEE, MMM d, ''yy"));
+
+set hive.local.time.zone=US/Hawaii;
+
+select from_unixtime(unix_timestamp('1400-11-08 01:53:11'));
+select from_unixtime(unix_timestamp('1800-11-08 01:53:11'));
+select from_unixtime(to_unix_timestamp('0000-00-00'));
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/udf5.q:68]: Nit: Add extra line at EOF.

File: ql/src/test/queries/clientpositive/udf_to_unix_timestamp.q [hunk 7]
```diff
@@ -18,8 +18,8 @@ SELECT
 FROM oneline_n0;
 
 SELECT
-  '2009 Mar 20 11:30:01 am',
-  to_unix_timestamp('2009 Mar 20 11:30:01 am', 'yyyy MMM dd h:mm:ss a')
+  '2009 Mar 20 11:30:01 AM',
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/udf_to_unix_timestamp.q:21]: Why is this change required? Lower case doesn't work?

File: ql/src/test/queries/clientpositive/vector_unix_timestamp.q [hunk 8]
```diff
@@ -1,6 +1,6 @@
 
 create table t (d string);
-insert into t values('2020-11-16 22:18:40 UTC');
+insert into t values('2020-11-16 22:18:40');
```
Comment (sankarh) [inline ql/src/test/queries/clientpositive/vector_unix_timestamp.q:3]: Why is it changed?

File: ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java (hunk 4 above)
Comment (ashish-kumar-sharma) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDFToUnixTimeStamp.java:182]: Done

File: ql/src/test/queries/clientpositive/udf5.q (hunk 5 above)
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/udf5.q:12]: Sure.

File: ql/src/test/queries/clientpositive/udf5.q (hunk 6 above)
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/udf5.q:68]: Done

File: ql/src/test/queries/clientpositive/udf_to_unix_timestamp.q (hunk 7 above)
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/udf_to_unix_timestamp.q:21]: simpleDateFromat and DateTimeformatter both used locale(https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html) to parse alphabet. By default system local in picked to in both the case. simpleDateFromat use Locale.getDefault(Locale.Category.FORMAT)) which is list of locale due to "AM" and "am" but not "aM" or "Am". Where as DateTimeformatter used only system default locale also DateTimeformatter by default in case Sensitive as well. If supports only "AM"/"am" which is not in our control. I Do align the point the date value should be case insensitive but I couldn't able to find any DateTimeFromatter implementation with in hive code. But still in order to make unix_timestamp backword compatible we need to make DateTimeformatter case insensitive.

File: ql/src/test/queries/clientpositive/vector_unix_timestamp.q (hunk 8 above)
Comment (ashish-kumar-sharma) [reply ql/src/test/queries/clientpositive/vector_unix_timestamp.q:3]: Since with introduction of DateTimefomatter we only parse value which fall under the pattern "yyyy-MM-dd HH:mm:ss" other then that we just drop. But after we have introduce the resolverstyle.strict it will strictly check for the pattern "yyyy-MM-dd HH:mm:ss" . Other then that it will fail and return run. since in function like unix_timestamp() we can provide the pattern to resolve the date/Time. So i have added both the test cases for better understanding.

**Comment by ashish-kumar-sharma** [comment]
@mattmccline-microsoft Could you please review the PR?

**Review by sankarh — APPROVED** [review]
